proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,726 @@
+
+     Without ``-XElidedTypes``, an underscore in a type is an error; the error message would suggest
+     either using a named wildcard to get a diagnostic or enabling ``-XElidedTypes`` to accept the
+     elision.
+
+     Each elision spelled ``__`` will cause GHC to print out a diagnostic warning explaining how
+     the elision was filled. Writing an elision as ``_`` suppresses the diagnostic, as does
+     ``-Wno-elided-types``.
+
+   - In expressions, a ``_`` is a part of the expression the author did not care to write. Currently, this means
+     that ``_`` will be replaced with ``error "elision at <line>:<col>"``. In this case, an error will be printed,
+     stating the inferred type of the ``_`` and suggesting to enable ``-XElidedExpressions`` if the user
+     wants to keep the ``error``\ing behavior. With ``-XElidedExpressions``, GHC will still warn, controlled
+     by ``-Werroring-elided-expressions``. (This case is an exception to the general rule that ``_`` prints
+     no diagnostic, as it seems willfully cruel not to print one here. Once GHC can infer proper expressions,
+     this might be changed.)
+
+     In the future, GHC may support the possibility of inferring expressions. An elided expression may then
+     be filled in, not with a call to ``error``, but a correct expression. For example, we might imagine
+     that ::
+
+       id :: a -> a
+       id x = _
+
+     has its elision filled in with the only possible non-bottom value, ``x``. Any such behavior will
+     have to be specified separately and would likely be guarded by an extension flag and possible diagnostic
+     output.
+
+2. Outside of patterns, treat an unbound identifier beginning with an underscore as a named wildcard. A named wildcard
+   induces GHC to print an error with the wildcard's type and a suggested value. This behavior is
+   controlled by the ``-XNamedWildCards`` extension.
+
+   - In a type signature, a named wildcard behaves much as one does today, though its kind will be printed
+     in the diagnostic along with the other information. Just like today, a suggestion will be included
+     to enable ``-XPartialTypeSignatures``.
+
+   - In a type outside a type signature (e.g., in an instance declaration or a data constructor type),
+     a named wildcard will induce a diagnostic including the wildcard's kind and any information GHC
+     can figure out about its value.
+
+   - In an expression, a named wildcard will induce a diagnostic including the
+     wildcard's type and any information GHC can figure out about its value,
+     including suggested replacements. In this way, an expression named
+     wildcard will behave like holes have.
+
+   In all cases, the diagnostic is suppressed by prefixing the identifier with only one underscore;
+   two or more underscores will induce the diagnostic.
+
+3. In the absence of ``-XNamedWildCards``, the use of a type variable that begins with an underscore will induce a warning
```

Comment:
Can we just deprecate the old behavior? I realize that takes time, but I think it would be worthwhile.